### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by Veganveins**
This PR attempts to test how well we are able to estimate initial state probabilities given a transition probability matrix.

I've included two first pass attempts to:

1. Test the `compute_steady_state` method

I'm having some trouble figuring out what makes sense to assert here. Currently I'm trying to assert that the output of `compute_steady_state` is a TensorVariable, but I think what we actually want to do is assert that the steady state probabilities have been properly solved for? 

2. Test the `gamma_0_estimation` method

Once I create a model I'm running into an error that 

```python
with test_model:
        ffbs = FFBSStep([S_rv])
-->  trace = pm.sample()

SamplingError: Initial evaluation of model at starting point failed!
```

I'm having a hard time interpreting that error / trying to debug it. Initially I thought it might be due to the way I initialized `p_0_rv` and `p_1_rv` but I'm not sure. Any suggestions on where I might be going wrong here / how I might write these tests a bit differently?

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -120,3 +122,11 @@ def test_multilogit_inv(test_input, test_output):
...
     res = res.eval()
     assert np.array_equal(res.round(2), test_output)
+
+
+def test_compute_steady_state():
+
+    P = tt.as_tensor_variable(np.arange(4).reshape(2, 2).astype(np.float32))
+
+    ss_probs = compute_steady_state(P)
+    assert isinstance(ss_probs, TensorVariable)
```
Comment (Veganveins) [inline tests/test_utils.py:132]: This isn't the way we want to be writing this test, but I'm a little stuck on coming up with an alternate approach. ss_probs looks something like this:

```python
In [6]: ss_probs
Out[6]: Solve{A_structure='general', lower=False, overwrite_A=False, overwrite_b=False}.0
```

and it has a lot of things that can be called. It seemed like maybe we could assert that the probabilities sum to 1 with something like

```python
ss_probs.sum()
Sum{acc_dtype=float64}.0
```

but the output of that doesn't seem to be a float, it's a TensorVariable itself

File: tests/test_estimation.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        # how well is this estimated?
+        pi_0_rv = pm.Dirichlet("pi_0", np.r_[1, 1])  # "flat" prior for initial states
+
+        S_rv = DiscreteMarkovChain("S_t", P_rv, pi_0_rv, shape=y_test.shape[0])
+
+        PoissonZeroProcess("Y_t", 150, S_rv, observed=y_test)
+
+    with test_model:
+        ffbs = FFBSStep([S_rv])
+        trace = pm.sample()
```
Comment (Veganveins) [inline tests/test_estimation.py:36]: This `pm.sample` step raises this error:

```python
SamplingError: Initial evaluation of model at starting point failed!
```

And I'm having a hard time figuring out exactly what's causing that error, any suggestions on where to start debugging this?

Comment (brandonwillard) [reply tests/test_estimation.py:36]: The initial point is available from `test_model.test_point`.  If you were to plug those values into the model's log-likelihood function (e.g. `test_model.logp(test_model.test_point)`) and get something like `-np.inf`, you would see an error like that.

From there, you can check the log-likelihoods for each term (e.g. `test_model.S_t.logp(test_model.test_point)`.

I'm guessing that the `S_t` values are causing the a `-np.inf` log-likelihood.  The zero/non-zero HMMs are _very_ unforgiving; if an observation is zero/non-zero and the corresponding state value in the initial value array is non-zero/zero, then you'll get `-np.inf`.  

In other words, if you set up a model that says there's a non-zero/zero value emitted from the zero/non-zero state's distribution, the log-likelihood is going to tell you that this situation is impossible (i.e. `-np.inf`).

File: tests/test_estimation.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+        P_tt = tt.stack([p_0_rv, p_1_rv])  # transition matrix
+        P_rv = pm.Deterministic("P_tt", tt.shape_padleft(P_tt))
+
+        # how well is this estimated?
+        pi_0_rv = pm.Dirichlet("pi_0", np.r_[1, 1])  # "flat" prior for initial states
+
+        S_rv = DiscreteMarkovChain("S_t", P_rv, pi_0_rv, shape=y_test.shape[0])
+
+        PoissonZeroProcess("Y_t", 150, S_rv, observed=y_test)
```
Comment (brandonwillard) [inline tests/test_estimation.py:32]: Since we only want to test the initial state estimation, let's simplify this model so that nothing else interferes with that.  Instead of using a Poisson, let's start by using a mixture with two `pm.Constant` distributions (for the simulated "true" data, as well).

If we use `pm.Constant`s at the same values as the state values (i.e. `0` and `1`), then we can effectively bypass the emission distributions and only deal with the state sequence and initial states.  The log-likelihoods end up being much simpler and easier to reason about on the fly.

File: tests/test_utils.py (hunk 1 above)
Comment (brandonwillard) [reply tests/test_utils.py:132]: You're dealing with symbolic variables in Aesara/Theano here.  When you want to compute values for the those symbolic operations, you need to compile their symbolic graphs using `theano.function`.  See the documentation [here](https://aesara.readthedocs.io/en/latest/introduction.html#sneak-peek).

File: tests/test_estimation.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        pi_0_rv = pm.Dirichlet("pi_0", np.r_[1, 1])  # "flat" prior for initial states
+
+        # take the transition matrix and the flat prior for initial states
+        # creates a sequence of states and it "chooses" the state and time t based
+        # on how the trans matrix P_rv is defined and
+        S_rv = DiscreteMarkovChain("S_t", P_rv, pi_0_rv, shape=y_test.shape[0])
+
+        # takes series of states
+        Y_t = SwitchingProcess("Y_t", [pm.Constant.dist(0), pm.Constant.dist(1)], S_rv)
+
```
Comment (Veganveins) [inline tests/test_estimation.py:39]: should I be referencing `Y_t` somewhere within the model context? currently flake8 is complaining that `local variable "Y_t" is imported but never used`

File: tests/test_estimation.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    # simulate fake data
+    sim_dat = np.random.binomial(1, 0.5, size=y_test.shape[0])
+
+    with test_model:
+
+        prior_predictive = pm.sample_prior_predictive(samples=100)
+        assert 0.49 < prior_predictive["Y_t"].mean() < 0.51
+
+        # use model to get a posterior distribution
+        trace = pm.sample()
```
Comment (Veganveins) [inline tests/test_estimation.py:49]: `trace = pm.sample()` here is failing with error `RuntimeError: Chain 0 failed.`

Is this the wrong way to be trying to get a posterior trace with `test_model`?

File: tests/test_estimation.py (hunk 4 above)
Comment (brandonwillard) [reply tests/test_estimation.py:39]: Not necessarily.  If we're not using it, we can get rid of that variable (not the `SwitchingProcess` call, though).

File: tests/test_estimation.py (hunk 5 above)
Comment (brandonwillard) [reply tests/test_estimation.py:49]: If the model above is only used for producing a set of sample observations&mdash;as it appears to be&mdash;then another copy of that model will be needed for the estimation.  That copy will need to set the `observed` keyword argument in the `Y_t` term.

In other words, this code is currently asking for posterior samples from a model that has no observations, and, thus, no likelihood(s).

Take a look at the simulation + estimation setup used in the example notebooks.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: tests/utils.py [hunk 6]
```diff
@@ -16,7 +16,10 @@ def simulate_poiszero_hmm(
         P_tt = tt.stack([p_0_rv, p_1_rv])
         P_rv = pm.Deterministic("P_tt", tt.shape_padleft(P_tt))
 
-        pi_0_tt = pm.Dirichlet("pi_0", pi_0_a)
+        if pi_0:
+            pi_0_tt = pm.Dirichlet("pi_0", pi_0)
+        else:
+            pi_0_tt = pm.Dirichlet("pi_0", pi_0_a)
```
Comment (Veganveins) [inline tests/utils.py:23]: @brandonwillard I think you helped me edit this `simulate_poiszero_hmm` function when I started working on this but I think I lost that change when I was trying to re-organize the commits. Does this if-else condition look right to you? I can't remember if this was exactly how we had it before

Comment (brandonwillard) [reply tests/utils.py:23]: I don't recall; you'll need to check the Git history or walk through whatever problems this is causing.

File: tests/test_estimation.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    p_1_rv = pm.Dirichlet("p_1", p_1_a)
+    P_tt = tt.stack([p_0_rv, p_1_rv])
+    P_rv = pm.Deterministic("P_tt", tt.shape_padleft(P_tt))
+    if pi_0 is not None:
+        pi_0_tt = tt.as_tensor(pi_0)
+    else:
+        pi_0_tt = pm.Dirichlet("pi_0", pi_0_a)
+    S_rv = DiscreteMarkovChain("S_t", P_rv, pi_0_tt, shape=N)
+    S_rv.tag.test_value = (observed > 0) * 1
+    return PoissonZeroProcess("Y_t", mu, S_rv, observed=observed)
```
Comment (brandonwillard) [inline tests/test_estimation.py:28]: As a simple test for initial probability estimation, I don't know if the zero-Poisson model is really the best choice.

If we're going to use a model with a Dirac emission distribution, we should use one that consists entirely of Dirac emissions; it's simpler and accomplishes the same thing.

Otherwise, we need to consider whether or not a non-Dirac model is a better choice for testing (e.g. mixture of normals).  Likewise for the choice of transition matrix.

File: tests/test_estimation.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def test_gamma_0_estimation():
+
+    np.random.seed(2032)
+    true_initial_states = np.array([0.5, 0.5])
```
[the remark this answers is not in the document]
Comment (Veganveins) [reply tests/test_estimation.py:34]: if I changed it to something like [.32, .68] would that be better? or are we thinking even more extreme than that, like [.02, .98] or something?

[the remark this answers is not in the document]
Comment (brandonwillard) [reply tests/test_estimation.py:34]: Let's try the latter.

File: tests/test_estimation.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    S_rv = DiscreteMarkovChain("S_t", P_rv, pi_0_tt, shape=N)
+    S_rv.tag.test_value = (observed > 0) * 1
+    return PoissonZeroProcess("Y_t", mu, S_rv, observed=observed)
+
+
+def test_gamma_0_estimation():
+
+    np.random.seed(2032)
+    true_initial_states = np.array([0.02, 0.98])
+
```
Comment (Veganveins) [inline tests/test_estimation.py:35]: seems like with true initial states set to [.02, .98] the initial state estimation fails. the error im seeing is

```python
E       AssertionError: 
E       Arrays are not almost equal to 1 decimals
E       
E       Mismatched elements: 2 / 2 (100%)
E       Max absolute difference: 0.25904421
E       Max relative difference: 12.95221072
E        x: array([0.3, 0.7])
E        y: array([0., 1.])
```
